src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs

```diff
@@ -29309,32 +29325,369 @@ class C<T>
+            comp.VerifyDiagnostics(
+                // (12,9): warning CS8762: Parameter 'y' may not have a null value when exiting with 'false'.
+                //         return y != null || z != null; // 1, 2
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y != null || z != null;").WithArguments("y", "false").WithLocation(12, 9),
+                // (12,9): warning CS8762: Parameter 'z' may not have a null value when exiting with 'false'.
+                //         return y != null || z != null; // 1, 2
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return y != null || z != null;").WithArguments("z", "false").WithLocation(12, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("TStruct?")]
+        [InlineData("TClass?")]
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_NotNullWhen(string type)
+        {
+            var source = @"
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
+class C<TStruct, TClass>
+    where TStruct : struct
+    where TClass : class
+{
+    static bool TryGetValue([NotNullWhen(true)] out TYPE y)
+    {
+        y = null;
+        if (y == null)
+        {
+            return true; // 1
+        }
+        return false;
+    }
+
+    static bool TryGetValue2([NotNullWhen(true)] out TYPE y)
+    {
+        y = null;
+        return y != null;
+    }
+
+    static bool TryGetValue3([NotNullWhen(false)] out TYPE y)
+    {
+        y = null;
+        return y == null;
+    }
+
+    static bool TryGetValue4([NotNull] TYPE x, [NotNullWhen(false)] out TYPE y)
+    {
+        y = null;
+        if (y != null)
+        {
+            return true;
+        }
+        return false; // 2
+    }
+}
+";
+            var comp = CreateNullableCompilation(new[] { NotNullAttributeDefinition, NotNullWhenAttributeDefinition, source.Replace("TYPE", type) });
+            comp.VerifyDiagnostics(
+                // (15,13): error CS8762: Parameter 'y' may not have a null value when exiting with 'true'.
+                //             return true; // 1
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return true;").WithArguments("y", "true").WithLocation(15, 13),
+                // (39,9): error CS8762: Parameter 'y' may not have a null value when exiting with 'false'.
+                //         return false; // 2
+                Diagnostic(ErrorCode.WRN_ParameterConditionallyDisallowsNull, "return false;").WithArguments("y", "false").WithLocation(39, 9)
+                );
+        }
+
+        [Theory]
+        [InlineData("TStruct?")]
+        [InlineData("TClass?")]
+        [WorkItem(39922, "https://github.com/dotnet/roslyn/issues/39922")]
+        public void EnforcedInMethodBody_NotNullWhen_Composition(string type)
```

Comment:
I can't find tests which demonstrate the expected behavior around composition like the following:

```cs
bool TryGetValue1(string key, [MaybeNullWhen(false)] out string value) => TryGetValue2(key, out value);
bool TryGetValue2(string key, [NotNullWhen(true)] out string? value) => TryGetValue1(key, out value);
```

Reply:
Added. Thanks

---
In reply to: [369852520](https://github.com/dotnet/roslyn/pull/41098#discussion_r369852520) [](ancestors = 369852520)